azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -303,17 +371,20 @@ def _block(self, op):
     def _unblock(self, op, error):
         """
         Unblock this stage after the connection is complete.  This also means
-        releasing all the queued up operations that we were waiting for the
-        connect operation to complete.
+        releasing all the operations that were queued up.
         """
         logger.info("{}({}): disabling block and releasing queued ops.".format(self.name, op.name))
         self.blocked = False
         logger.info(
             "{}({}): processing {} items in queue".format(self.name, op.name, self.queue.qsize())
         )
-        # loop through our queue and release all the blocked operations
-        while not self.queue.empty():
-            op_to_release = self.queue.get_nowait()
+        # Loop through our queue and release all the blocked operations
+        # Put a new Queue in self.queue because releasing ops might put them back in the
+        # queue, especially if there's a ConnectOperation in the list of ops to release
+        old_queue = self.queue
+        self.queue = queue.Queue()
+        while not old_queue.empty():
+            op_to_release = old_queue.get_nowait()
             if error:
                 # if we're unblocking the queue because something (like a connect operation) failed,
                 # then we fail all of the blocked operations with the same error.
```

Comment:
Is this desired? Shouldn't we retry here? Or will that be part of a future PR?

Reply:
retry needs to be handled in a different stage and is part of a different PR.

---
In reply to: [311715312](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r311715312) [](ancestors = 311715312)